src/a2dp-lhdc.c

```diff
@@ -0,0 +1,516 @@
+	if (configuration->max_bitrate == LHDC_MAX_BITRATE_400K) {
+		return LHDCBT_QUALITY_LOW;
+	} else if (configuration->max_bitrate == LHDC_MAX_BITRATE_500K) {
+		return LHDCBT_QUALITY_MID;
+	} else {
+		return LHDCBT_QUALITY_HIGH;
+	}
+}
+
+void *a2dp_lhdc_enc_thread(struct ba_transport_pcm *t_pcm) {
+
+	pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
+	pthread_cleanup_push(PTHREAD_CLEANUP(ba_transport_pcm_thread_cleanup), t_pcm);
+
+	struct ba_transport *t = t_pcm->t;
+	struct io_poll io = { .timeout = -1 };
+
+	const a2dp_lhdc_v3_t *configuration = &t->a2dp.configuration.lhdc_v3;
+
+	HANDLE_LHDC_BT handle;
+	if ((handle = lhdcBT_get_handle(get_version(configuration))) == NULL) {
+		error("Couldn't get LHDC handle: %s", strerror(errno));
+		goto fail_open_lhdc;
+	}
+
+	pthread_cleanup_push(PTHREAD_CLEANUP(lhdcBT_free_handle), handle);
+
+	const unsigned int channels = t_pcm->channels;
+	const unsigned int samplerate = t_pcm->sampling;
+	const unsigned int bitdepth = get_bit_depth(configuration);
+
+	lhdcBT_set_hasMinBitrateLimit(handle, configuration->min_bitrate);
+	lhdcBT_set_max_bitrate(handle, get_max_bitrate(configuration));
+
+	struct {
+		uint8_t seq_num;
+		uint8_t latency:2;
+		uint8_t frames:6;
+	} *lhdc_media_header;
+
+	rtp_header_t *rtp_header;
+
+	if (lhdcBT_init_encoder(handle, samplerate, bitdepth, config.lhdc_eqmid,
+			configuration->ch_split_mode > LHDC_CH_SPLIT_MODE_NONE, 0, t->mtu_write - sizeof(*lhdc_media_header) - sizeof(*rtp_header) + sizeof(rtp_header->csrc),
+			get_encoder_interval(configuration)) == -1) {
+		error("Couldn't initialize LHDC encoder");
+		goto fail_init;
+	}
+
+	const size_t lhdc_ch_samples = lhdcBT_get_block_Size(handle);
+	const size_t lhdc_pcm_samples = lhdc_ch_samples * channels;
+
+	ffb_t bt = { 0 };
+	ffb_t pcm = { 0 };
+	pthread_cleanup_push(PTHREAD_CLEANUP(ffb_free), &bt);
+	pthread_cleanup_push(PTHREAD_CLEANUP(ffb_free), &pcm);
+
+	int32_t *pcm_ch1 = malloc(lhdc_ch_samples * sizeof(int32_t));
+	int32_t *pcm_ch2 = malloc(lhdc_ch_samples * sizeof(int32_t));
+	pthread_cleanup_push(PTHREAD_CLEANUP(free), pcm_ch1);
+	pthread_cleanup_push(PTHREAD_CLEANUP(free), pcm_ch2);
+
+	if (ffb_init_int32_t(&pcm, lhdc_pcm_samples) == -1 ||
+			ffb_init_uint8_t(&bt, t->mtu_write) == -1 ||
+			pcm_ch1 == NULL || pcm_ch2 == NULL) {
+		error("Couldn't create data buffers: %s", strerror(errno));
+		goto fail_ffb;
+	}
+
+
+
+	/* initialize RTP headers and get anchor for payload */
+	uint8_t *rtp_payload = rtp_a2dp_init(bt.data, &rtp_header,
+			(void **)&lhdc_media_header, sizeof(*lhdc_media_header));
+
+	struct rtp_state rtp = { .synced = false };
+	/* RTP clock frequency equal to audio samplerate */
+	rtp_state_init(&rtp, samplerate, samplerate);
+
+	uint8_t seq_num = 0;
+
+	debug_transport_pcm_thread_loop(t_pcm, "START");
+	for (ba_transport_pcm_state_set_running(t_pcm);;) {
+
+		switch (io_poll_and_read_pcm(&io, t_pcm, &pcm)) {
+		case -1:
+			if (errno == ESTALE) {
+				ffb_rewind(&pcm);
```

Comment:
Is there any function for flushing encoder internal buffers? If yes, it should be used here. Please, see other A2DP codecs for reference.

Reply:
Nope, there are none.

Reply:
Is there any internal state of the encoder? The use case looks like this: application sends PCM and then it calls "drop" call which should drop all previously written PCM samples (not necessarily transmitted over the Bluetooth link). Then, the application sends new samples and these new samples should be transmitted without any residual samples from the previous write. Most encoders keep some internal state which is based on some samples history. So, in order to have a clean start, this state needs to be flushed. Otherwise, in case of writing silence after the "drop" call and encoding it, the decoder will decode some non-zero samples. Some encoders do have flush API, but some need a trick to do it (e.g.: lc3plus).

Reply:
There is a circular buffer, which accumulates samples. But there is no API to drop samples from it.

Reply:
OK, so some workaround will have to be applied here.